lib/ReplaceOpenCLBuiltinPass.cpp

```diff
@@ -660,6 +675,191 @@ Type *ReplaceOpenCLBuiltinPass::GetPairStruct(Type *type) {
   return new_struct;
 }
 
+bool ReplaceOpenCLBuiltinPass::replaceWaitGroupEvents(Function &F) {
+  /* Simple implementation for wait_group_events to avoid dealing with the event
+   * list:
+   *
+   * void wait_group_events(int num_events, event_t *event_list) {
+   *   barrier(CLK_LOCAL_MEM_FENCE);
+   * }
+   *
+   */
+
+  enum {
+    CLK_LOCAL_MEM_FENCE = 0x01,
+    CLK_GLOBAL_MEM_FENCE = 0x02,
+    CLK_IMAGE_MEM_FENCE = 0x04
+  };
+
+  return replaceCallsWithValue(F, [](CallInst *CI) {
+    IRBuilder<> Builder(CI);
+
+    const auto ConstantScopeWorkgroup = Builder.getInt32(spv::ScopeWorkgroup);
+    const auto MemorySemanticsWorkgroup = BinaryOperator::Create(
+        Instruction::Shl, Builder.getInt32(CLK_LOCAL_MEM_FENCE),
+        Builder.getInt32(clz(spv::MemorySemanticsWorkgroupMemoryMask) -
+                         clz(CLK_LOCAL_MEM_FENCE)),
+        "", CI);
+
+    return clspv::InsertSPIRVOp(CI, spv::OpControlBarrier,
+                                {Attribute::NoDuplicate, Attribute::Convergent},
+                                Builder.getVoidTy(),
+                                {ConstantScopeWorkgroup, ConstantScopeWorkgroup,
+                                 MemorySemanticsWorkgroup});
+  });
+}
+
+Value *ReplaceOpenCLBuiltinPass::replaceAsyncWorkGroupCopies(
+    Module &M, CallInst *CI, Value *Dst, Value *Src, Value *NumGentypes,
+    Value *Stride, Value *Event) {
+  /*
+   * event_t *async_work_group_strided_copy(T *dst, T *src, size_t num_gentypes,
+   *                                        size_t stride, event_t event) {
+   *   size_t start_id = ((get_local_id(2) * get_local_size(1))
+   *                     + get_local_id(1)) * get_local_size(0)
+   *                     + get_local_id(0);
+   *   size_t incr = get_local_size(0) * get_local_size(1) * get_local_size(2);
+   *   for (size_t it = start_id; it < num_gentypes; it += incr) {
+   *     dst[it] = src[it * stride];
+   *   }
+   *   return event;
+   * }
+   */
+
+  /* BB:
+   *    before
+   *    async_work_group_strided_copy
+   *    after
+   *
+   * ================================
+   *
+   * BB:
+   *    before
+   *    start_id = f(get_local_ids, get_local_sizes)
+   *    incr = g(get_local_sizes)
+   *    br CmpBB
+   *
+   * CmpBB:
+   *    it = PHI(start_id, it)
+   *    cmp = it < NumGentypes
+   *    condBr cmp, LoopBB, ExitBB
+   *
+   * LoopBB:
+   *    dstI = dst[it]
+   *    srcI = src[it * stride]
+   *    OpCopyMemory dstI, srcI
+   *    it += incr
+   *    br CmpBB
+   *
+   * ExitBB:
+   *    after
+   */
+
+  IRBuilder<> Builder(CI);
+
+  auto Cst0 = Builder.getInt32(0);
+  auto Cst1 = Builder.getInt32(1);
+  auto Cst2 = Builder.getInt32(2);
+
+  // get_local_id({0, 1, 2});
+  GlobalVariable *GVId = M.getGlobalVariable("__spirv_LocalInvocationId");
```

Comment:
I think it's time to move "__spirv_LocalInvocationId" and "__spirv_WorkgroupSize" into lib/Constants.h and replace their uses with functions.

Reply:
Separately, have you tested this against a kernel that doesn't already use `get_local_id()`? I'm not sure the local invocation id variable will be automatically generated in all cases.

Reply:
No, I did not test this against a kernel without a call to `get_local_id`. I will do that

Reply:
I fix this with this commit 2e1889c
It adds tests as well for it